crates/web3/src/schema.rs

```diff
@@ -0,0 +1,565 @@
+//! using Borsh serialization. It supports both unsigned and signed transactions
+//! with configurable schema validation.
+//!
+//! ## When to Use This Module
+//!
+//! Use this module when:
+//! - Working with language bindings where Rust generics are not available
+//! - Building dynamic systems that need runtime schema validation
+//! - Creating Web3 interfaces for languages like JavaScript, Python, or Go
+//! - The transaction types are not known at compile time
+//! - You need to serialize transactions based on external schema definitions
+//!
+//! ## Comparison with `rust` Module
+//!
+//! This crate provides two different approaches for transaction building and serialization:
+//!
+//! ### `rust` Module (Compile-time Generics)
+//! - **Use case**: Pure Rust applications with compile-time type safety
+//! - **Type safety**: Full compile-time validation using Rust's type system
+//! - **Performance**: Zero-cost abstractions with compile-time optimizations
+//! - **Flexibility**: Requires specific generic type parameters (`Spec`, `ChainHash`, etc.)
+//! - **Target audience**: Rust developers building native applications
+//!
+//! ### `schema` Module (Runtime Schema)
+//! - **Use case**: Language bindings and dynamic systems
+//! - **Type safety**: Runtime validation using JSON schema definitions
+//! - **Performance**: Runtime serialization overhead but more flexible
+//! - **Flexibility**: Works with any JSON-based transaction structure
+//! - **Target audience**: Multi-language environments and dynamic systems
+//!
+//! The `schema` module is specifically designed for scenarios where Rust's
+//! compile-time generics cannot be used, such as FFI bindings to other programming
+//! languages or systems that need to work with dynamically defined transaction formats.
+
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+use sov_universal_wallet::schema::{RollupRoots, Schema};
+
+pub use serde_json::json;
+
+/// Errors that can occur during schema-based serialization operations.
+#[derive(thiserror::Error, Debug)]
+pub enum SerializerError {
+    /// Error occurred during JSON to Borsh conversion using the schema.
+    #[error("Borsh serialization error: {0}")]
+    JsonToBorsh(#[source] sov_universal_wallet::schema::SchemaError),
+    /// Error occurred during JSON serialization or deserialization.
+    #[error("JSON serialization error: {0}")]
+    JsonSerialization(#[from] serde_json::Error),
+    /// The provided schema is invalid or malformed.
+    #[error("Invalid schema: {0}")]
+    InvalidSchema(#[source] serde_json::Error),
+    #[error("Failed to calculate chain hash: {0}")]
+    ChainHash(#[source] sov_universal_wallet::schema::SchemaError),
+    /// Error occurred during HTTP request to fetch schema from URL.
+    #[error("HTTP request error: {0}")]
+    HttpRequest(#[from] reqwest::Error),
+    #[error("schema object from http response was invalid")]
```

Comment:
```suggestion
    #[error("HTTP schema request returned an invalid response")]
```
"Schema object invalid" is a bit confusing when there's an `InvalidSchema` error right above, for actually invalid schema objects